tests/vm_delete_protection_test.go

```diff
@@ -72,24 +79,192 @@ var _ = Describe("VM delete protection", func() {
+			g.Expect(updatedNs.DeletionTimestamp).ToNot(BeNil())
+		}, env.ShortTimeout(), time.Second).Should(Succeed())
+
+		Expect(apiClient.Get(ctx, client.ObjectKey{Name: vm.Name, Namespace: vm.Namespace}, vm)).ToNot(HaveOccurred())
+
+		Consistently(func() error {
+			controllerRevision := &appsv1.ControllerRevision{}
+			return apiClient.Get(ctx, client.ObjectKey{Name: controllerRevisionName, Namespace: vm.Namespace}, controllerRevision)
+		}, 30*time.Second, time.Second).Should(Succeed(), "controllerRevision should not be deleted")
+	})
+
+	It("should be able to delete the namespace if the VM delete protection is disabled", func() {
+		namespaceName := fmt.Sprintf("test-ns-%v", rand.String(5))
+		ns := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: namespaceName,
+			},
+		}
+		Expect(apiClient.Create(ctx, ns)).To(Succeed())
+		DeferCleanup(func() {
+			_ = apiClient.Delete(ctx, ns)
+		})
+
+		vm = createVMWithDeleteProtection("true", ns.Name)
+
+		startVM(vm.Name, vm.Namespace, vmReadyTimeout)
+
+		_ = apiClient.Delete(ctx, ns)
+
+		// Just check the namespace is in Terminating state (has DeletionTimestamp)
+		Eventually(func(g Gomega) {
+			updatedNs := &corev1.Namespace{}
+			err := apiClient.Get(ctx, client.ObjectKey{Name: namespaceName}, updatedNs)
+			g.Expect(err).ToNot(HaveOccurred())
+			g.Expect(updatedNs.DeletionTimestamp).ToNot(BeNil())
+		}, env.ShortTimeout(), time.Second).Should(Succeed())
+
+		// Disable the delete protection
+		Eventually(func() error {
+			Expect(apiClient.Get(ctx, client.ObjectKey{Name: vm.Name, Namespace: vm.Namespace}, vm)).ToNot(HaveOccurred())
+			vm.Labels[deleteProtectionLabel] = "False"
+			return apiClient.Update(ctx, vm)
+		}, env.ShortTimeout(), time.Second).ToNot(HaveOccurred())
+
+		// Check the namespace is deleted
+		Eventually(func() error {
+			updatedNs := &corev1.Namespace{}
+			return apiClient.Get(ctx, client.ObjectKey{Name: namespaceName}, updatedNs)
+		}, env.ShortTimeout(), time.Second).Should(MatchError(errors.IsNotFound, "errors.IsNotFound"))
+	})
 })
 
-func createVMWithDeleteProtection(protected string) *kubevirtv1.VirtualMachine {
-	return createVMWithLabels(map[string]string{deleteProtectionLabel: protected})
+func createVMWithDeleteProtection(protected string, namespace string) *kubevirtv1.VirtualMachine {
+	return createVMWithLabels(map[string]string{deleteProtectionLabel: protected}, namespace)
 }
 
-func createVMWithLabels(labels map[string]string) *kubevirtv1.VirtualMachine {
+func createVMWithLabels(labels map[string]string, namespace string) *kubevirtv1.VirtualMachine {
 	vmName := fmt.Sprintf("testvmi-%v", rand.String(10))
-	vmi := NewMinimalVMIWithNS(strategy.GetNamespace(), vmName)
+	vmi := NewMinimalVMIWithNS(namespace, vmName)
+	vmi.Spec.Volumes = []kubevirtv1.Volume{
+		{
+			Name: "containerdisk",
+			VolumeSource: kubevirtv1.VolumeSource{
+				ContainerDisk: &kubevirtv1.ContainerDiskSource{
+					Image: "quay.io/containerdisks/debian:latest",
```

Comment:
Why is this required?

Reply:
Yes, I was wondering if that's why you added this. The CR creation doesn't rely on the guest OS, just the VM control loop in `virt-controller`.